src/perm.c

```diff
@@ -9,85 +9,26 @@
-            mpz_setbit(perm, bit);
-            mismatches--;
-        }
-    }
-
-    mpz_clears(binom, curr_ordinal, NULL);
-}
-
-void get_random_permutation(mpz_t perm, int mismatches, size_t subkey_length,
-                            gmp_randstate_t randstate) {
-    mpz_t ordinal, binom;
-    mpz_inits(ordinal, binom, NULL);
-
-    mpz_bin_uiui(binom, subkey_length, mismatches);
-
-    mpz_urandomm(ordinal, randstate, binom);
-    decode_ordinal(perm, ordinal, mismatches, subkey_length);
-
-    mpz_clears(ordinal, binom, NULL);
-}
-
-void get_benchmark_permutation(mpz_t perm, int mismatches, size_t subkey_length,
-                               gmp_randstate_t randstate, int numcores) {
-    mpz_t ordinal, binom, rank, cores;
-    mpz_inits(ordinal, binom, rank, NULL);
-    mpz_init_set_ui(cores, numcores);
-
-    mpz_bin_uiui(binom, subkey_length, mismatches);
-
-    // Choose a random rank from 0 to numcores - 1
-    mpz_urandomm(rank, randstate, cores);
-
-    mpz_mul_ui(rank, rank, 2);
-    mpz_add_ui(rank, rank, 1);
-    mpz_mul(ordinal, binom, rank);
-
-    mpz_tdiv_q_ui(ordinal, ordinal, numcores * 2);
-
-    decode_ordinal(perm, ordinal, mismatches, subkey_length);
-
-    mpz_clears(ordinal, binom, rank, cores, NULL);
-}
+void decodeOrdinal(mpz_t perm, const mpz_t ordinal, int mismatches, size_t subkey_length);
 
 /// Assigns the first possible permutation for a given # of mismatches.
 /// \param perm A pre-allocated mpz_t to fill the permutation to.
 /// \param mismatches The hamming distance that you want to base the permutation on.
-void assign_first_permutation(mpz_t perm, int mismatches) {
-    // Set perm to first key
-    // Equivalent to: (perm << mismatches) - 1
-    mpz_set_ui(perm, 1);
-    mpz_mul_2exp(perm, perm, mismatches);
-    mpz_sub_ui(perm, perm, 1);
-}
+void assignFirstPermutation(mpz_t perm, int mismatches);
 
 /// Assigns the first possible permutation for a given # of mismatches and key size
 /// \param perm A pre-allocated mpz_t to fill the permutation to.
 /// \param mismatches The hamming distance that you want to base the permutation on.
 /// \param subkey_length How big the relevant key is in # of bits.
-void assign_last_permutation(mpz_t perm, int mismatches, size_t subkey_length) {
-    // First set the value to the first permutation.
-    assign_first_permutation(perm, mismatches);
-    // Equivalent to: perm << (subkey_length - mismatches)
-    // E.g. if subkey_length = 256 and mismatches = 5
-    // Then we want to shift left 256 - 5 = 251 times.
-    mpz_mul_2exp(perm, perm, subkey_length - mismatches);
-}
+void assignLastPermutation(mpz_t perm, int mismatches, size_t subkey_length);
+
+void getRandomPermutation(mpz_t perm, int mismatches, size_t subkey_length,
+                          gmp_randstate_t randstate);
 
-void get_random_seed(unsigned char *key, size_t key_size, gmp_randstate_t randstate) {
+void getBenchmarkPermutation(mpz_t perm, int mismatches, size_t subkey_length,
```

Comment:
Why'd this funciton show up

Reply:
That's probably from me re-arranging so that the internal *.c functions definitions are after the *.h functions, kind of like putting public functions before private.